File: Reverse_Bits.cpp
/*190. Reverse Bits Add to List
Description  Submission  Solutions
Total Accepted: 94428
Total Submissions: 320606
Difficulty: Easy
Contributors: Admin
Reverse bits of a given 32 bits unsigned integer.

For example, given input 43261596 (represented in binary as 00000010100101000001111010011100), return 964176192 (represented in binary as 00111001011110000010100101000000).

Follow up:
If this function is called many times, how would you optimize it?

Related problem: Reverse Integer

Credits:
Special thanks to @ts for adding this problem and creating all test cases.

Hide Company Tags Apple Airbnb
Hide Tags Bit Manipulation
Hide Similar Problems (E) Number of 1 Bits
*/




/*each time add one bit, if the original bit is 0, then OR | operation should be 0
if the original bit is 1, then add 1
start from the end of the bit */

/////////////////////////////////////////////////////////////////////////////////////
//C++
class Solution {
public:
    uint32_t reverseBits(uint32_t n) {
        // uint32_t r;
        // r = 0;
        // for(int i = 0; i < 32; i++)
        // {
        //     r <<= 1;
        //     r += n & 1;
        //     n >>= 1;
        // }
        // return r;

        uint32_t ret = 0;
        for(int i = 0; i <32; ++i, n >>= 1){
            ret <<= 1;
            ret |= n & 1;
        }

        return ret;
    }
};





/*
Solution 1 and Solution 2 take similar time

Bit manipulation

Solution 1:
each time
    num <<= 1;
    num = num | n&1;
    n >>= 1;

Solution 2:  2 ms / 600 test
use unsigned right shift
and signed left shift
2 digits as a group
    eg: take odd digits, right shift 1 digits
        take even digits, left shift 1 digits
4 digits as a group
8 digits as a group
16 digits as a group
32 digits as a group

Follow up:
if this function is called multiple times
divide an int into 4 bytes,
reverse each byte then combine into an int
for each byte, we could use cache to improve performance
 */

/////////////////////////////////////////////////////////////////////////////////////
//Java
//Solution 1
public class Solution {
    // you need treat n as an unsigned value
    public int reverseBits(int n) {
        int ret = 0;
        //need to check it 32times
        for(int i = 0; i < 32; ++i, n >>= 1){
            ret <<= 1;
            ret |= n & 1; //num = num | n&1;
        }

        return ret;
    }
}



//Solution 2:
public class Solution {
    // you need treat n as an unsigned value
    public int reverseBits(int n) {
        //unsigned right shift
        //signed left shift
        n = ((n & 0xAAAAAAAA) >>> 1) | ((n & 0x55555555) << 1);
        n = ((n & 0xCCCCCCCC) >>> 2) | ((n & 0x33333333) << 2);
        n = ((n & 0xf0f0f0f0) >>> 4) | ((n & 0x0f0f0f0f) << 4);
        n = ((n & 0xff00ff00) >>> 8) | ((n & 0x00ff00ff) << 8);
        n = ((n & 0xffff0000) >>> 16) | ((n & 0x0000ffff) << 16);
        return n;
    }
}



//Follow up
public class Solution {
    // you need treat n as an unsigned value
    private Map<Byte, Integer> cache = new HashMap<>();

    public int reverseBits(int n) {
        byte[] bytes = new byte[4];

        //convert n into 4 byte[]
        for(int i = 0; i < 4; ++i){
            bytes[i] = (byte)((n >>> 8*i) & 0xff);
        }

        //reverse each byte
        int num = 0; //return number
        for(int i = 0; i < 4; ++i){
            num += reverseByte(bytes[i]);
            if(i < 3){
                num <<= 8;
            }
        }
        return num;
    }

    public int reverseByte(byte b){
        Integer val = cache.get(b);
        //if we have already reverse this 8 bits
        if(val != null) return val;

        val = 0;
        //reverse bit
        for(int i = 0; i < 8; ++i, b >>= 1){
            val <<= 1;
            val |= (b&1);
        }
        return val;
    }
}

File: Reverse_Bits.swift
/*
Reverse bits of a given 32 bits unsigned integer.

Note:

Note that in some languages such as Java, there is no unsigned integer type. In this case, both input and output will be given as a signed integer type. They should not affect your implementation, as the integer's internal binary representation is the same, whether it is signed or unsigned.
In Java, the compiler represents the signed integers using 2's complement notation. Therefore, in Example 2 above, the input represents the signed integer -3 and the output represents the signed integer -1073741825.
Follow up:

If this function is called many times, how would you optimize it?

 

Example 1:

Input: n = 00000010100101000001111010011100
Output:    964176192 (00111001011110000010100101000000)
Explanation: The input binary string 00000010100101000001111010011100 represents the unsigned integer 43261596, so return 964176192 which its binary representation is 00111001011110000010100101000000.
Example 2:

Input: n = 11111111111111111111111111111101
Output:   3221225471 (10111111111111111111111111111111)
Explanation: The input binary string 11111111111111111111111111111101 represents the unsigned integer 4294967293, so return 3221225471 which its binary representation is 10111111111111111111111111111111.
 

Constraints:

The input must be a binary string of length 32
*/

/*
Solution 2
bit shift

            b1 b2
shift -> b1 0,   0 b2
shift -> 0 b1,  b2 0
combine -> b2 b1

32bit 0000 0000 0000 0000 0000 0000 0000 0000
		f	f	 0	   0 |  f	f	 0	   0
		0	0	 f	   f |  0	0 	 f	   f
		f	0  | f	   0 |  f	0  | f	   0
		0	f  | 0	   f |  0	f  | 0	   f
	  1100 1100 1100 1100 1100 1100 1100 1100
	    c	c    c	   c    c	c    c	   c
	  0011 0011 0011 0011 0011 0011 0011 0011
	    3	3    3	   3    3	3    3	   3
	  1010 1010 1010 1010 1010 1010 1010 1010
	    a	a    a	   a    a	a    a	   a
	  0101 0101 0101 0101 0101 0101 0101 0101
	    5	5    5	   5    5	5    5	   5

Time Complexity: O(1)
Space Complexity: O(1)
*/
class Solution {
    func reverseBits(_ n: Int) -> Int {
        var n = (n >> 16) | (n << 16)
        n = ((n & 0xff00ff00) >> 8) | ((n & 0x00ff00ff) << 8)
        n = ((n & 0xf0f0f0f0) >> 4) | ((n & 0x0f0f0f0f) << 4)
        n = ((n & 0xcccccccc) >> 2) | ((n & 0x33333333) << 2)
        n = ((n & 0xaaaaaaaa) >> 1) | ((n & 0x55555555) << 1)
        return n
    }
}

/*
Solution 1
Bit by bit

for each bit, add it to res with power

Time Complexity: O(1)
Space Complexity: O(1)
*/
class Solution {
    func reverseBits(_ n: Int) -> Int {
        var n = n
        var res = 0
        
        var power = 31
        while n != 0 {
            res += (n&1) << power
            n = n >> 1
            power -= 1
        }
        
        return res
    }
}